Source/WebCore/platform/graphics/ImageSource.cpp

```diff
@@ -431,6 +448,30 @@ void ImageSource::stopAsyncDecodingQueue()
     LOG(Images, "ImageSource::%s - %p - url: %s [decoding has been stopped]", __FUNCTION__, this, sourceURL().string().utf8().data());
 }
 
+RefPtr<NativeImage> ImageSource::createFrameImageFromCachedDecodedImageAtIndex(size_t index, SubsamplingLevel subsamplingLevel, const DecodingOptions& decodingOptions)
+{
+    if (index >= m_frames.size())
+        return nullptr;
+
+    ImageFrame& frame = m_frames[index];
+
+    if (!frame.hasCachedDecodedImageCompatibleWithOptions(subsamplingLevel, decodingOptions))
+        return nullptr;
+
+    return std::exchange(frame.m_cachedDecodedImage, { })->createNativeImage();
+}
+
+bool ImageSource::setFrameImageFromCachedDecodedImageAtIndex(size_t index, SubsamplingLevel subsamplingLevel, const DecodingOptions& decodingOptions)
+{
+    auto image = createFrameImageFromCachedDecodedImageAtIndex(index, subsamplingLevel, decodingOptions);
+    if (!image)
+        return false;
+
+    cacheNativeImageAtIndex(WTFMove(image), index, subsamplingLevel, decodingOptions);
```

Comment:
It is a little bit strange that we get a NativeImage out of the ImageFrame (by calling createFrameImageFromCachedDecodedImageAtIndex ()) and then we set it back (by calling cacheNativeImageAtIndex()). Should not this be done in one function instead?